File: ios/enve/Screens/Sources/ServerProbe.swift
import Foundation

struct DetectedServer: Identifiable {
    let providerType: ProviderType
    let displayName: String
    let normalizedURL: String
    let oidcEnabled: Bool
    let recommendedAuth: UnifiedAuthMethod

    var id: String { providerType.rawValue + "|" + normalizedURL }
}

enum ServerProbeOutcome {

    case identified(DetectedServer)

    case oidcIssuerOnly(normalizedURL: String)

    case unknown(normalizedURL: String)

    case unreachable
}

enum ServerProbe {

    static func detect(rawURL: String) async -> ServerProbeOutcome {
        let bases = candidateBases(from: rawURL)
        guard !bases.isEmpty else { return .unreachable }

        var reachableBase: String?
        for base in bases {
            if let detected = await identify(base) {
                return .identified(detected)
            }
            if let issuer = await oidcIssuer(at: base) {
                return .oidcIssuerOnly(normalizedURL: issuer)
            }
            if reachableBase == nil, await isReachable(base) {
                reachableBase = base
            }
        }

        if let reachableBase { return .unknown(normalizedURL: reachableBase) }
        return .unreachable
    }

    private static func identify(_ base: String) async -> DetectedServer? {

        async let abs = fingerprintAudiobookshelf(base)
        async let grimmory = fingerprintGrimmory(base)
        async let bookOrbit = fingerprintBookOrbit(base)
        async let silo = fingerprintSilo(base)
        async let komga = fingerprintKomga(base)
        async let plex = fingerprintPlex(base)
        async let jellyfinEmby = fingerprintJellyfinEmby(base)
        async let storyteller = fingerprintStoryteller(base)
        async let kavita = fingerprintKavita(base)
        async let opds = fingerprintOPDS(base)

        let ordered: [DetectedServer?] = [
            await abs, await grimmory, await bookOrbit, await silo, await komga,
            await plex, await jellyfinEmby, await storyteller, await kavita, await opds,
        ]
        return ordered.compactMap { $0 }.first
    }

    private struct ABSStatus: Decodable {
        let app: String?
        let authMethods: [String]?
    }

    private static func fingerprintAudiobookshelf(_ base: String) async -> DetectedServer? {
        guard let (resp, data) = await get(base, "/status"), resp.statusCode == 200,
            let status = try? JSONDecoder().decode(ABSStatus.self, from: data),
            status.app?.lowercased() == "audiobookshelf"
        else { return nil }
        let oidc = (status.authMethods ?? []).contains { $0.lowercased().contains("openid") }
        return DetectedServer(
            providerType: .audiobookshelf,
            displayName: "Audiobookshelf",
            normalizedURL: base,
            oidcEnabled: oidc,
            recommendedAuth: oidc ? .oidc : .usernamePassword
        )
    }

    private struct GrimmorySettings: Decodable {
        let oidcEnabled: Bool
    }

    private static func fingerprintGrimmory(_ base: String) async -> DetectedServer? {
        guard let (resp, data) = await get(base, "/api/v1/public-settings"), resp.statusCode == 200,
            let settings = try? JSONDecoder().decode(GrimmorySettings.self, from: data)
        else { return nil }
        return DetectedServer(
            providerType: .booklore,
            displayName: "Grimmory",
            normalizedURL: base,
            oidcEnabled: settings.oidcEnabled,
            recommendedAuth: settings.oidcEnabled ? .oidc : .usernamePassword
        )
    }

    private struct BookOrbitProvider: Decodable {
        let slug: String?
    }

    private static func fingerprintBookOrbit(_ base: String) async -> DetectedServer? {
        guard let (resp, data) = await get(base, "/api/v1/app-settings/oidc/providers/public"),
            resp.statusCode == 200,
            let providers = try? JSONDecoder().decode([BookOrbitProvider].self, from: data)
        else { return nil }
        let oidc = !providers.isEmpty
        return DetectedServer(
            providerType: .bookOrbit,
            displayName: "BookOrbit",
            normalizedURL: base,
            oidcEnabled: oidc,
            recommendedAuth: oidc ? .oidc : .usernamePassword
        )
    }

    private struct SiloHealth: Decodable {
        let status: String?
        let serverID: String?

        enum CodingKeys: String, CodingKey {
            case status
            case serverID = "server_id"
        }
    }

    private static func fingerprintSilo(_ base: String) async -> DetectedServer? {

        guard let (resp, data) = await get(base, "/api/v1/health"), resp.statusCode == 200,
            let health = try? JSONDecoder().decode(SiloHealth.self, from: data),
            health.status?.lowercased() == "ok",
            health.serverID?.isEmpty == false
        else { return nil }
        return DetectedServer(
            providerType: .silo,
            displayName: "Silo",
            normalizedURL: base,
            oidcEnabled: false,
            recommendedAuth: .usernamePassword
        )
    }

    private struct KomgaClaim: Decodable {
        let isClaimed: Bool
    }

    private static func fingerprintKomga(_ base: String) async -> DetectedServer? {
        guard let (resp, data) = await get(base, "/api/v1/claim"), resp.statusCode == 200,
            (try? JSONDecoder().decode(KomgaClaim.self, from: data)) != nil
        else { return nil }
        let oidc = await komgaHasOIDC(base)
        return DetectedServer(
            providerType: .komga,
            displayName: "Komga",
            normalizedURL: base,
            oidcEnabled: oidc,
            recommendedAuth: oidc ? .oidc : .usernamePassword
        )
    }

    private static func komgaHasOIDC(_ base: String) async -> Bool {
        guard let (resp, data) = await get(base, "/api/v1/oauth2/providers"), resp.statusCode == 200 else { return false }
        if let array = try? JSONSerialization.jsonObject(with: data) as? [Any] { return !array.isEmpty }
        if let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] { return !object.isEmpty }
        return false
    }

    private static func fingerprintPlex(_ base: String) async -> DetectedServer? {
        guard let (resp, data) = await get(base, "/identity", accept: "application/json"),
            resp.statusCode == 200,
            let body = String(data: data, encoding: .utf8)?.lowercased(),
            body.contains("machineidentifier")
        else { return nil }
        return DetectedServer(
            providerType: .plex,
            displayName: "Plex",
            normalizedURL: base,
            oidcEnabled: false,
            recommendedAuth: .usernamePassword
        )
    }

    private struct PublicSystemInfo: Decodable {
        let ProductName: String?
        let ServerName: String?
        let Version: String?
        let Id: String?
    }

    private static func fingerprintJellyfinEmby(_ base: String) async -> DetectedServer? {
        let lowerBase = base.lowercased()
        let apiBases =
            lowerBase.hasSuffix("/emby")
            ? [(url: base, isEmbyAPIBase: true)]
            : [(url: base, isEmbyAPIBase: false), (url: base + "/emby", isEmbyAPIBase: true)]

        for apiBase in apiBases {
            guard let (resp, data) = await get(apiBase.url, "/System/Info/Public"),
                resp.statusCode == 200
            else { continue }

            let info = try? JSONDecoder().decode(PublicSystemInfo.self, from: data)
            var product = (info?.ProductName ?? "").lowercased()
            if product.isEmpty, let text = String(data: data, encoding: .utf8)?.lowercased() {
                product = text
            }
            if product.contains("jellyfin") {
                return DetectedServer(
                    providerType: .jellyfin,
                    displayName: "Jellyfin",
                    normalizedURL: apiBase.url,
                    oidcEnabled: false,
                    recommendedAuth: .usernamePassword
                )
            }
            if product.contains("emby") || (apiBase.isEmbyAPIBase && isValidPublicSystemInfo(info)) {
                return DetectedServer(
                    providerType: .emby,
                    displayName: "Emby",
                    normalizedURL: apiBase.url,
                    oidcEnabled: false,
                    recommendedAuth: .usernamePassword
                )
            }
        }
        return nil
    }

    private static func isValidPublicSystemInfo(_ info: PublicSystemInfo?) -> Bool {
        guard let info else { return false }
        return info.ServerName?.isEmpty == false
            && info.Version?.isEmpty == false
            && info.Id?.isEmpty == false
    }

    private static func fingerprintStoryteller(_ base: String) async -> DetectedServer? {

        if let (resp, data) = await get(base, "/api/v2/auth/providers"), resp.statusCode == 200,
            let providers = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            providers["credentials"] != nil
        {
            let sso = providers.keys.contains { $0.lowercased() != "credentials" }
            return DetectedServer(
                providerType: .storyteller,
                displayName: "Storyteller",
                normalizedURL: base,
                oidcEnabled: sso,
                recommendedAuth: sso ? .webLogin : .usernamePassword
            )
        }
        guard let (resp, data) = await get(base, "/api"), resp.statusCode == 200,
            let body = String(data: data, encoding: .utf8)?.lowercased(),
            body.contains("\"hello\"") || body.contains("storyteller")
        else { return nil }
        return DetectedServer(
            providerType: .storyteller,
            displayName: "Storyteller",
            normalizedURL: base,
            oidcEnabled: false,
            recommendedAuth: .usernamePassword
        )
    }

    private static func fingerprintKavita(_ base: String) async -> DetectedServer? {
        guard let (resp, data) = await get(base, "/api/health"), resp.statusCode == 200 else { return nil }
        let body = (String(data: data, encoding: .utf8) ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard body == "ok" || body == "healthy" else { return nil }
        return DetectedServer(
            providerType: .kavita,
            displayName: "Kavita",
            normalizedURL: base,
            oidcEnabled: false,
            recommendedAuth: .usernamePassword
        )
    }

    private static func fingerprintOPDS(_ base: String) async -> DetectedServer? {
        guard let (resp, data) = await get(base, ""), resp.statusCode == 200 else { return nil }
        let contentType = (resp.value(forHTTPHeaderField: "Content-Type") ?? "").lowercased()
        let body = String(data: data.prefix(2048), encoding: .utf8)?.lowercased() ?? ""
        let isOPDS =
            contentType.contains("atom+xml") || contentType.contains("opds")
            || (body.contains("<feed") && body.contains("www.w3.org/2005/atom"))
        guard isOPDS else { return nil }
        return DetectedServer(
            providerType: .opds,
            displayName: "OPDS",
            normalizedURL: base,
            oidcEnabled: false,
            recommendedAuth: .usernamePassword
        )
    }

    private struct OIDCDiscovery: Decodable {
        let issuer: String
        let authorization_endpoint: String
    }

    private static func oidcIssuer(at base: String) async -> String? {
        guard let (resp, data) = await get(base, "/.well-known/openid-configuration"),
            resp.statusCode == 200,
            (try? JSONDecoder().decode(OIDCDiscovery.self, from: data)) != nil
        else { return nil }
        return base
    }

    private static func isReachable(_ base: String) async -> Bool {
        await get(base, "") != nil
    }

    private static func get(
        _ base: String,
        _ path: String,
        accept: String? = nil,
        timeout: TimeInterval = 6
    ) async -> (HTTPURLResponse, Data)? {
        guard let url = URL(string: base + path) else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = timeout
        if let accept { request.setValue(accept, forHTTPHeaderField: "Accept") }
        do {
            let (data, response) = try await InsecureURLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else { return nil }
            return (http, data)
        } catch {
            return nil
        }
    }

    private static func candidateBases(from raw: String) -> [String] {
        var value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        while value.hasSuffix("/") { value.removeLast() }
        guard !value.isEmpty else { return [] }
        let lower = value.lowercased()
        if lower.hasPrefix("http://") || lower.hasPrefix("https://") {
            return [value, alternateScheme(for: value)].compactMap { $0 }
        }

        guard let preferred = ServerURLNormalizer.normalize(rawURL: value, providerType: .emby)?.absoluteString else {
            return []
        }
        return [preferred, alternateScheme(for: preferred)].compactMap { $0 }
    }

    private static func alternateScheme(for value: String) -> String? {
        if value.lowercased().hasPrefix("https://") {
            return "http://" + String(value.dropFirst("https://".count))
        }
        if value.lowercased().hasPrefix("http://") {
            return "https://" + String(value.dropFirst("http://".count))
        }
        return nil
    }
}

File: android/core/src/main/java/com/enve/core/data/provider/ServerProbe.kt
package com.enve.core.data.provider

import com.enve.core.data.model.BookSource
import com.enve.core.data.model.ConnectionAuthMode
import com.enve.core.data.remote.security.PrivateNetworkTrust
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.security.SecureRandom
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager

data class DetectedServer(
    val source: BookSource,
    val displayName: String,
    val normalizedUrl: String,
    val oidcEnabled: Boolean,
    val recommendedAuth: ConnectionAuthMode,
)

sealed interface ServerProbeOutcome {
    data class Identified(val server: DetectedServer) : ServerProbeOutcome
    data class OidcIssuerOnly(val normalizedUrl: String) : ServerProbeOutcome
    data class Unknown(val normalizedUrl: String) : ServerProbeOutcome
    data object Unreachable : ServerProbeOutcome
}

object ServerProbe {

    private val AUTO_SSO_SOURCES = setOf(
        BookSource.AUDIOBOOKSHELF, BookSource.GRIMMORY, BookSource.STORYTELLER,
    )

    suspend fun detect(rawUrl: String): ServerProbeOutcome {
        val bases = candidateBases(rawUrl)
        if (bases.isEmpty()) return ServerProbeOutcome.Unreachable

        var reachable: String? = null
        for (base in bases) {
            identify(base)?.let { return ServerProbeOutcome.Identified(it) }
            oidcIssuer(base)?.let { return ServerProbeOutcome.OidcIssuerOnly(it) }
            if (reachable == null && isReachable(base)) reachable = base
        }
        return reachable?.let { ServerProbeOutcome.Unknown(it) } ?: ServerProbeOutcome.Unreachable
    }

    private suspend fun identify(base: String): DetectedServer? = coroutineScope {

        val probes = listOf(
            async { fingerprintAudiobookshelf(base) },
            async { fingerprintGrimmory(base) },
            async { fingerprintBookOrbit(base) },
            async { fingerprintKomga(base) },
            async { fingerprintPlex(base) },
            async { fingerprintJellyfinEmby(base) },
            async { fingerprintStoryteller(base) },
            async { fingerprintSilo(base) },
            async { fingerprintKavita(base) },
            async { fingerprintOpds(base) },
        )
        probes.firstNotNullOfOrNull { it.await() }
    }

    private fun ssoOrPassword(source: BookSource, oidc: Boolean): ConnectionAuthMode =
        if (oidc && source in AUTO_SSO_SOURCES) ConnectionAuthMode.SSO else ConnectionAuthMode.USERNAME_PASSWORD

    private suspend fun fingerprintAudiobookshelf(base: String): DetectedServer? {
        val r = get(base, "/status") ?: return null
        if (r.code != 200) return null
        val json = r.body.toJsonObjectOrNull() ?: return null
        if (json.optString("app").lowercase() != "audiobookshelf") return null
        val methods = json.optJSONArray("authMethods")
        val oidc = (0 until (methods?.length() ?: 0)).any {
            methods!!.optString(it).lowercase().contains("openid")
        }
        return DetectedServer(BookSource.AUDIOBOOKSHELF, "Audiobookshelf", base, oidc,
            ssoOrPassword(BookSource.AUDIOBOOKSHELF, oidc))
    }

    private suspend fun fingerprintGrimmory(base: String): DetectedServer? {
        val r = get(base, "/api/v1/public-settings") ?: return null
        if (r.code != 200) return null
        val json = r.body.toJsonObjectOrNull() ?: return null
        if (!json.has("oidcEnabled")) return null
        val oidc = json.optBoolean("oidcEnabled")
        return DetectedServer(BookSource.GRIMMORY, "Grimmory", base, oidc,
            ssoOrPassword(BookSource.GRIMMORY, oidc))
    }

    private suspend fun fingerprintBookOrbit(base: String): DetectedServer? {
        val r = get(base, "/api/v1/app-settings/oidc/providers/public") ?: return null
        if (r.code != 200) return null
        val arr = r.body.toJsonArrayOrNull() ?: return null
        val oidc = arr.length() > 0
        return DetectedServer(BookSource.BOOKORBIT, "BookOrbit", base, oidc,
            ssoOrPassword(BookSource.BOOKORBIT, oidc))
    }

    private suspend fun fingerprintKomga(base: String): DetectedServer? {
        val r = get(base, "/api/v1/claim") ?: return null
        if (r.code != 200) return null
        val json = r.body.toJsonObjectOrNull() ?: return null
        if (!json.has("isClaimed")) return null
        return DetectedServer(BookSource.KOMGA, "Komga", base, komgaHasOidc(base),
            ConnectionAuthMode.USERNAME_PASSWORD)
    }

    private suspend fun komgaHasOidc(base: String): Boolean {
        val r = get(base, "/api/v1/oauth2/providers") ?: return false
        if (r.code != 200) return false
        r.body.toJsonArrayOrNull()?.let { return it.length() > 0 }
        r.body.toJsonObjectOrNull()?.let { return it.length() > 0 }
        return false
    }

    private suspend fun fingerprintPlex(base: String): DetectedServer? {
        val r = get(base, "/identity", accept = "application/json") ?: return null
        if (r.code != 200 || !r.body.lowercase().contains("machineidentifier")) return null
        return DetectedServer(BookSource.PLEX, "Plex", base, false, ConnectionAuthMode.USERNAME_PASSWORD)
    }

    private suspend fun fingerprintJellyfinEmby(base: String): DetectedServer? {
        val r = get(base, "/System/Info/Public") ?: return null
        if (r.code != 200) return null
        val product = r.body.toJsonObjectOrNull()?.optString("ProductName")?.lowercase()
            ?: r.body.lowercase()
        return when {
            product.contains("jellyfin") ->
                DetectedServer(BookSource.JELLYFIN, "Jellyfin", base, false, ConnectionAuthMode.USERNAME_PASSWORD)
            product.contains("emby") ->
                DetectedServer(BookSource.EMBY, "Emby", base, false, ConnectionAuthMode.USERNAME_PASSWORD)
            else -> null
        }
    }

    private suspend fun fingerprintStoryteller(base: String): DetectedServer? {

        val providers = get(base, "/api/v2/auth/providers")
        if (providers != null && providers.code == 200) {
            val obj = providers.body.toJsonObjectOrNull()
            if (obj != null && obj.has("credentials")) {
                val sso = obj.keys().asSequence().any { it.lowercase() != "credentials" }
                return DetectedServer(BookSource.STORYTELLER, "Storyteller", base, sso,
                    ssoOrPassword(BookSource.STORYTELLER, sso))
            }
        }
        val r = get(base, "/api") ?: return null
        if (r.code != 200) return null
        val lower = r.body.lowercase()
        if (!lower.contains("\"hello\"") && !lower.contains("storyteller")) return null
        return DetectedServer(BookSource.STORYTELLER, "Storyteller", base, false,
            ConnectionAuthMode.USERNAME_PASSWORD)
    }

    private suspend fun fingerprintSilo(base: String): DetectedServer? {

        val r = get(base, "/api/v1/health") ?: return null
        if (r.code != 200) return null
        val json = r.body.toJsonObjectOrNull() ?: return null
        if (json.optString("status") != "ok" || json.optString("server_id").isBlank()) return null
        return DetectedServer(BookSource.SILO, "Silo", base, false, ConnectionAuthMode.USERNAME_PASSWORD)
    }

    private suspend fun fingerprintKavita(base: String): DetectedServer? {
        val r = get(base, "/api/health") ?: return null
        if (r.code != 200) return null
        val trimmed = r.body.trim().lowercase()
        if (trimmed != "ok" && trimmed != "healthy") return null
        return DetectedServer(BookSource.KAVITA, "Kavita", base, false, ConnectionAuthMode.USERNAME_PASSWORD)
    }

    private suspend fun fingerprintOpds(base: String): DetectedServer? {
        val r = get(base, "") ?: return null
        if (r.code != 200) return null
        val contentType = (r.contentType ?: "").lowercase()
        val snippet = r.body.take(2048).lowercase()
        val isOpds = contentType.contains("atom+xml") || contentType.contains("opds") ||
            (snippet.contains("<feed") && snippet.contains("www.w3.org/2005/atom"))
        if (!isOpds) return null
        return DetectedServer(BookSource.OPDS, "OPDS", base, false, ConnectionAuthMode.USERNAME_PASSWORD)
    }

    private suspend fun oidcIssuer(base: String): String? {
        val r = get(base, "/.well-known/openid-configuration") ?: return null
        if (r.code != 200) return null
        val json = r.body.toJsonObjectOrNull() ?: return null
        if (!json.has("issuer") || !json.has("authorization_endpoint")) return null
        return base
    }

    private data class HttpResult(val code: Int, val body: String, val contentType: String?)

    private suspend fun isReachable(base: String): Boolean = get(base, "") != null

    private suspend fun get(base: String, path: String, accept: String? = null): HttpResult? =
        withContext(Dispatchers.IO) {
            val builder = runCatching { Request.Builder().url(base + path).get() }.getOrNull()
                ?: return@withContext null
            if (accept != null) builder.header("Accept", accept)
            try {
                client.newCall(builder.build()).execute().use { resp ->
                    HttpResult(resp.code, resp.body?.string().orEmpty(), resp.header("Content-Type"))
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                null
            }
        }

    private fun candidateBases(raw: String): List<String> {
        val value = raw.trim().trimEnd('/')
        if (value.isEmpty()) return emptyList()
        val lower = value.lowercase()
        return if (lower.startsWith("http://") || lower.startsWith("https://")) listOf(value)
        else listOf("https://$value", "http://$value")
    }

    private fun String.toJsonObjectOrNull(): JSONObject? = runCatching { JSONObject(this) }.getOrNull()
    private fun String.toJsonArrayOrNull(): JSONArray? = runCatching { JSONArray(this) }.getOrNull()

    private val client: OkHttpClient by lazy {
        val trustManager = PrivateNetworkTrust.buildTrustManager()
        val sslContext = SSLContext.getInstance("TLS").apply {
            init(null, arrayOf<TrustManager>(trustManager), SecureRandom())
        }
        OkHttpClient.Builder()
            .sslSocketFactory(sslContext.socketFactory, trustManager)
            .hostnameVerifier(PrivateNetworkTrust.buildHostnameVerifier())
            .connectTimeout(6, TimeUnit.SECONDS)
            .readTimeout(6, TimeUnit.SECONDS)
            .build()
    }
}
